File: Hashgard/Back/Hubs/ChatHub.cs
using Hashgard.Back.Hubs.Abstract;
using Hashgard.Back.Models;
using Hashgard.Back.Services;
using System;
using System.Threading.Tasks;

namespace Hashgard.Back.Hubs
{
    public class ChatHub : BaseHub<IChatHubClient>
    {
        private readonly IChatService _chatService;

        public ChatHub(IChatService chatService, IUserService userService)
            : base(userService, "Chat")
        {
            _chatService = chatService;
        }

        protected override Task AfterListenAsync(User user, string groupName, string connectionId)
        {
            return Clients.OthersInGroup(groupName).UserJoined(user);
        }
        

        public Task PostMessage(string tokenString, int categoryId, string text)
        {
            return ForLoggedUser(tokenString, async user =>
            {
                var message = await _chatService.PostMessageAsync(categoryId, text, user.Id);
                await Clients.OthersInGroup(GetGroupName(categoryId)).NewMessage(message);
            });
        }

        public Task PostEvent(string tokenString, int categoryId, ChatEventType eventType, string text)
        {
            return ForLoggedUser(tokenString, async user =>
            {
                var message = await _chatService.PostEventAsync(categoryId, eventType, user.Id);
                await Clients.OthersInGroup(GetGroupName(categoryId)).NewMessage(message);
            });
        }

        public Task ViewMessage(string tokenString, int messageId)
        {
            return ForLoggedUser(tokenString, async user =>
            {
                var view = await _chatService.ViewAsync(messageId, user.Id);
                await Clients.OthersInGroup(GetGroupName(view.ChatMessage.CategoryId)).MessageViewed(view);
            });
        }
    }

    public interface IChatHubClient : IHubClient
    {
        Task UserJoined(User user);
        Task NewMessage(ChatMessage message);
        Task MessageViewed(ChatView view);
    }
}

File: Hashgard/Front/src/app/services/chat/chat.hub.ts
import { BaseHub } from "../base-hub";
import { Injectable, EventEmitter } from "@angular/core";
import { User } from "../../models/user";
import { ChatMessageView } from "../../models/chat-message-view";
import { ChatMessage } from "../../models/chat-message";

@Injectable({
    providedIn: 'root'
})
export class ChatHub extends BaseHub {

    public userJoined: EventEmitter<User> = new EventEmitter();
    public newMessage: EventEmitter<ChatMessage> = new EventEmitter();
    public messageViewed: EventEmitter<ChatMessageView> = new EventEmitter();

    constructor() {
        super("chat");

        this.hubConnection.on("UserJoined", (user: User) => 
            this.userJoined.emit(user));

        this.hubConnection.on("NewMessage", (message: ChatMessage) => 
            this.newMessage.emit(message));

        this.hubConnection.on("MessageViewed", (view: ChatMessageView) =>
            this.messageViewed.emit(view));
    }

    public postMessage(categoryId: number, text: string) {
        this.hubConnection.invoke("PostMessage", this.userToken, categoryId, text);
    }

    public postEvent(categoryId: number, eventType: string) {
        this.hubConnection.invoke("PostEvent", this.userToken, categoryId, eventType);
    }

    public viewMessage(categoryId: number, messageId: number) {
        this.hubConnection.invoke("ViewMessage", this.userToken, categoryId, messageId);
    }
}
